api_app/services/cost_service.py

```diff
@@ -97,6 +125,34 @@ def __get_shared_services_costs(self, granularity, query_result_dict, shared_ser
             shared_services_costs.append(shared_service_cost_item)
         return shared_services_costs
 
+    def __get_workspace_services_costs(self, granularity, query_result_dict,
+                                       workspace_services_repo: WorkspaceServiceRepository,
+                                       user_resource_repo: UserResourceRepository, workspace_id: str):
+        workspace_services_costs = []
+        workspace_services_list = workspace_services_repo.get_active_workspace_services_for_workspace(workspace_id)
+        for workspace_service in workspace_services_list:
+
+            workspace_service_cost_item = WorkspaceServiceCostItem(**dict(
+                id=workspace_service.id,
+                name=self.__get_resource_name(workspace_service),
+                costs=self.__extract_cost_rows(granularity, query_result_dict, CostService.TRE_WORKSPACE_SERVICE_ID_TAG,
+                                               workspace_service.id),
+                user_resources=[]
+            ))
+
+            user_resources_list = user_resource_repo.get_user_resources_for_workspace_service(workspace_id,
+                                                                                              workspace_service.id)
+            for user_resource in user_resources_list:
+                user_resource_cost_item = CostItem(**dict(
+                    id=user_resource.id,
+                    name=self.__get_resource_name(user_resource),
+                    costs=self.__extract_cost_rows(granularity, query_result_dict, CostService.TRE_USER_RESOURCE_ID_TAG,
+                                                   user_resource.id)
+                ))
+                workspace_service_cost_item.user_resources.append(user_resource_cost_item)
+            workspace_services_costs.append(workspace_service_cost_item)
```

Comment:
I would extract into a method returning the workspace services list, in it a call to a method returning the resources list, and use list comprehension as much a possible

Reply:
refactored.
thanks @SharonHart , it saved many lines of code.